common/placer_heap.cc

```diff
@@ -841,7 +841,31 @@ class HeAPPlacer
         int ripup_radius = 2;
         int total_iters = 0;
         int total_iters_noreset = 0;
-        while (!remaining.empty()) {
+
+        bool need_legalize = require_validity;
+
+        /* First spread without legality checking, then check at the end! */
```

Comment:
I would ideally like a bit more investigation as to why this is needed, as I can't see an obvious reason.  It _could_ be the issue we've discussed before about how ripping up a cell can turn a legal placement illegal but I'm not sure if legaliser should ever be hitting that case.

Reply:
I agree. I think I'll finish getting the counter/RAM designs generating a useful DCP and then circle back and investigate this issue.

Reply:
I think I have a replicating test case where the current site router fails to find a solution, but a solution is possible.  Looks like a bug in the backtrace algorithm.  Debugging now, hopefully have a fix soon.

Reply:
Bug had to do with unrolling bad solutions, fix incoming.

Reply:
Fix is in https://github.com/YosysHQ/nextpnr/pull/645/commits/2a56539162e5d25e064a5940db75e8eb571ae917